jumpscale/sals/process/__init__.py

```diff
@@ -84,595 +141,833 @@ def execute(
     except Exception as e:
         raise j.exceptions.RuntimeError(e)
 
 
-def get_pid_by_port(port):
-    """Returns pids of the process that is listening on the given port
+def get_pid_by_port(port, ipv6=False, udp=False):
+    """Returns the PID of the process that is listening on the given port
 
-    Arguments:
-        port (int) -- port number
+    Args:
+        port (int): Port number to lookup for.
+        ipv6 (bool, optional): Whether to search the connections that using ipv6 instead of ipv4. Defaults to False.
+        udp (bool, optional): Whether to search the connections for UDP port instead of TCP. Defaults to False.
 
     Returns:
-        int -- pid of process that listen on that port
+        (int or None): Process ID that listen on that port
     """
 
-    process = get_process_by_port(port)
-    if process is None:
-        return []
-    return process.pid
+    process = get_process_by_port(port, ipv6=ipv6, udp=udp)
+    if process:
+        return process.pid
 
 
-def kill_process_by_name(name, sig=signal.SIGTERM.value, match_predicate=None):
-    """Kill all processes for a given command
+def kill_process_by_name(process_name, sig=signal.SIGTERM, match_predicate=None, timeout=5, sure_kill=False):
+    """Kill all processes that match 'process_name'.
 
-    Arguments:
-        name {str} -- Name of the command that started the process(s)
+    Args:
+        process_name (str): The target process name.
+        sig (signal, optional): See signal module constants. Defaults to signal.SIGKILL
+        match_predicate (callable, optional): Function that does matching between\
+            found processes and the targeted process, the function should accept\
+            two arguments and return a boolean. Defaults to None.
+        timeout (int, optional): How long to wait for a process to terminate (seconds) before raise exception\
+            or, if sure_kill=True, send a SIGKILL. Defaults to 5.
+        sure_kill (bool, optional): Whether to fallback to SIGKILL if the timeout exceeded for the terminate operation or not. Defaults to False.
 
-    Keyword Arguments:
-        sig {bool} -- os signal to send to the process(s) (default: {signal.SIGTERM.value})
-        match_predicate {callable} -- function that does matching between
-            found processes and the targested process, the function should accept
-            two arguments and return a boolean (default: {None})
+    Returns:
+        None or list of int: represents the IDs of the processes remaning alive.
     """
+    pids = get_pids(process_name, match_predicate=match_predicate)
+    failed_processes = []
+    for pid in pids:
+        try:
+            kill(pid, sig, timeout=timeout, sure_kill=sure_kill)
+        except (j.exceptions.Runtime, j.exceptions.Permission):
+            failed_processes.append(pid)
+    return failed_processes or None  # return None if the failed list is empty
```

Comment:
Can be empty list, no need for `None`?

Reply:
The same as a previous replay on similar comments :D